### ISSUE TITLE
Add tolerances for weight constraints

### ISSUE THREAD

**Issue by fontanf** [renault-iaa/challenge-roadef-2022#43]
Hi,

Several times, I got this kind of errors from the checker
```
Constraint W1;the weight of the stacks exceed the maximal authorized weight of the truck;truck id = PXXXXXXXXX;stacks weight = 25000,00;max authorized weigth of the truck = 25000
```
even though I try to do all computations as safe as possible.
This happens because weights are doubles. And just the order in which doubles are added could change the last digits of the final result.

It never happened to me for the axle weight constraints, but I guess it could also happen for these constraints.

Could it be possible to add a small tolerance, like 1e-5, in the checker to decrease the risk of this happening?

**Comment by renault-iaa** [comment]
Hello,

The maximal weight of 25000 kg is really a super max. We cannot add tolerances, you should really be below this weight limit.

**Comment by fontanf** [comment]
Hi,

Thank you for your answer, but I'm a bit surprised. Let me insist to be sure that my request is clear.

Here is a C++ example summing 5 weights appearing in instance `B/CL`:

```cpp
#include <limits>
#include <iostream>

int main(int argc, char *argv[])
{
    double w_0090016100_26102022000001 = 458.400;
    double w_0090016100_26102022000003 = 837.200;
    double w_0090016100_26102022000005 = 297.650;
    double w_0090016100_26102022000006 = 304.800;
    double w_0090016100_26102022000007 = 893.100;

    double res_1 = w_0090016100_26102022000007 + w_0090016100_26102022000003 + w_0090016100_26102022000001 + w_0090016100_26102022000006 + w_0090016100_26102022000005;
    double res_2 = w_0090016100_26102022000005 + w_0090016100_26102022000006 + w_0090016100_26102022000001 + w_0090016100_26102022000003 + w_0090016100_26102022000007;

    std::cout.precision(std::numeric_limits<double>::max_digits10);
    std::cout << "res_1 " << res_1 << std::endl;
    std::cout << "res_2 " << res_2 << std::endl;
    return 0;
}
```

These weights are summed all together in `res_1` and in `res_2`, but not in the same order. I print the result with full precision. Here is what I get on my machine:

```
res_1 2791.1500000000005
res_2 2791.1500000000001
```

Which means that if the capacity is `2791.1500000000001`, depending on the order in which the weights are summed, the solution might be considered feasible or not.

My request is just to consider the capacity in the checker as `25000.00001` to reduce the risk of this happening. And the same for the maximum weights on the axles.

Florian

**Comment by renault-iaa** [comment]
Hello,

The max weight will always be an integer. To make sure the max weight constraint is satisfied, you may satisfy the constraint ceil(sum of stacks' weights) <= truck's max weight

**Comment by fontanf** [comment]
It doesn't change anything if the capacity is integer or not. Here is another more explicit example. The capacity is set to `19000`. Depending on the order in which the total weight is computed, the solution `{3, 5, 6, 6, 11}` can be feasible or not:

```cpp
#include <limits>
#include <iostream>
#include <vector>

int main(int argc, char *argv[])
{
    std::cout.precision(std::numeric_limits<double>::max_digits10);

    int capacity = 19000;
    std::vector<double> weights = {458.400, 837.200, 297.650, 304.800, 893.100};
    std::vector<int> solution = {3, 5, 6, 6, 11};

    double res_1 = 0.0;
    for (int pos: {0, 1, 2, 3, 4})
        res_1 += solution[pos] * weights[pos];
    std::cout << "res_1 " << res_1 << std::endl;
    std::cout << "res_1 <= capacity " << (res_1 <= capacity) << std::endl;

    double res_2 = 0.0;
    for (int pos: {4, 3, 2, 1, 0})
        res_2 += solution[pos] * weights[pos];
    std::cout << "res_2 " << res_2 << std::endl;
    std::cout << "res_2 <= capacity " << (res_2 <= capacity) << std::endl;
    return 0;
}
```

Here is the output I get:
```
res_1 19000
res_1 <= capacity 1
res_2 19000.000000000004
res_2 <= capacity 0
```

**Comment by renault-iaa** [comment]
Hello,

Your precision issue seems associated with your code C++, we did not encounter such precision problems with java. In addition, all the weights have the precision of grams in our Information Systems. So managing weigths with 3 decimals should be enough. Otherwise you can also multiply all the weights by 1000.0 to obtain weights in grams and then manage these weigths as integers.

**Comment by fontanf** [comment]
The double type works the same way in C++ and in Java. The error message I got with the checker that I posted in the first message shows that it might happen in java too, and in particular with the checker
```
Constraint W1;the weight of the stacks exceed the maximal authorized weight of the truck;truck id = PXXXXXXXXX;stacks weight = 25000,00;max authorized weigth of the truck = 25000
```
I don't have the corresponding solution files anymore unfortunately, but that solution would have been feasible if weights and capacities had been multiplied by 1000.

The checker should allow a tolerance, or multiply the weights and capacities by 1000 and convert them to integer as you suggested. But this second fix doesn't solve the same issue that could arise with the axle weights constraints.

If the precision of the weights is to the gram, then adding a tolerance of 1e-5, or even 1e-4, won't make it possible to generate a solution overusing the capacity.
